### PR TITLE
API: libpacemaker: Move crm_rule functionality to libpacemaker

### PR DESCRIPTION
**Pull request by nrwahl2**
Closes T58

### CODE REVIEW

File: lib/pacemaker/pcmk_rule.c [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+ * \param[in]     input     The CIB XML to check (if \c NULL, use current CIB)
+ * \param[in]     date      Check whether the rule is in effect at this date and
+ *                          time (if \c NULL, use current date and time)
+ * \param[in]     rule_ids  The IDs of the rules to check
+ *
+ * \return Standard Pacemaker return code
+ */
+int
+pcmk__check_rules(pcmk__output_t *out, xmlNodePtr input, crm_time_t *date,
+                  const char **rule_ids)
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_rule.c:240]: I think it makes more sense for the API functions should handle just one rule at a time, instead of processing the whole list at once as `crm_rule` does. That way, we get to keep the return code for every rule we check, instead of just "success or the last non-success return code."

For this draft, I wrote it to accept a list of rule IDs like the existing logic in `crm_rule`. That way the output structure doesn't change. This can easily be updated to take just one rule ID though.

File: cts/cli/regression.rules.exp [hunk 2]
```diff
@@ -11,23 +11,17 @@ crm_rule: --check requires use of --rule=
 =#=#=#= End test: crm_rule given no rule to check - Incorrect usage (64) =#=#=#=
 * Passed: crm_rule       - crm_rule given no rule to check
 =#=#=#= Begin test: crm_rule given invalid input XML =#=#=#=
-(log_xmllib_err@xml.c:891) 	error: XML Error: Entity: line 1: parser error : Start tag expected, '<' not found
```
Comment (nrwahl2) [inline cts/cli/regression.rules.exp:14]: Do we need this? As mentioned in the commit message, we can get this (minus the `"@xml.c:891"`) if add verbosity, but we no longer get it by default.

File: tools/crm_rule.c [hunk 3]
```diff
@@ -87,175 +82,6 @@ mode_cb(const gchar *option_name, const gchar *optarg, gpointer data, GError **e
...
-        }
-    }
-
-    match = getXpathResult(xpathObj, 0);
-
-    /* We should have ensured both of these pass with the xpath query above, but
-     * double checking can't hurt.
-     */
-    CRM_ASSERT(match != NULL);
-    CRM_ASSERT(find_expression_type(match) == time_expr);
```
Comment (nrwahl2) [inline tools/crm_rule.c:248]: `time_expr` was not defined anywhere so I dropped it in when moving to libpacemaker

**Comment by nrwahl2** [comment]
@clumens Can you review when you get a chance?

File: lib/pacemaker/pcmk_rule.c [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+    freeXpathObject(xpath_obj);
+    return rc;
+}
+
+/*!
+ * \internal
+ * \brief Initialize the cluster working set for checking rules
+ *
+ * Make our own copies of the CIB XML and date/time object, if they're not
+ * \c NULL. This way we don't have to take ownership of the objects passed via
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_rule.c:159]: seems simpler for callers to deal with this way

File: cts/cli/regression.rules.exp (hunk 2 above)
Comment (nrwahl2) [reply cts/cli/regression.rules.exp:14]: This is actually what caused the CI build to fail.

`build-all-voting` succeeded, because `cts-cli` doesn't produce these `(log_xmllib_err@xml.c:891)` messages, which I removed from the test output.

However, `build-rpms` failed. The `(log_xmllib_err@xml.c:891)` messages DO appear in the output when running `make -C rpm rpm`.

I don't know what could be causing this discrepancy.

File: include/pacemaker.h [hunk 5]
```diff
@@ -186,6 +186,20 @@ int pcmk_list_nodes(xmlNodePtr *xml, char *node_types);
...
+ *
+ * \param[in,out] xml       The destination for the result, as an XML tree
+ * \param[in]     input     The CIB XML to check (if \c NULL, use current CIB)
+ * \param[in]     date      Check whether the rule is in effect at this date and
+ *                          time (if \c NULL, use current date and time)
+ * \param[in]     rule_ids  The IDs of the rules to check
+ *
+ * \return Standard Pacemaker return code
+ */
+int pcmk_check_rules(xmlNodePtr *xml, xmlNodePtr input, crm_time_t *date,
```
Comment (nrwahl2) [inline include/pacemaker.h:200]: Not sure why these are getting documented in the header file when we've typically put doxygen comments in the code, but I went ahead and followed the pattern.

File: cts/cli/regression.rules.exp (hunk 2 above)
Comment (nrwahl2) [reply cts/cli/regression.rules.exp]: Okay, here's the deal... Initially, I ran `cts-cli -t rules` after writing the code in this PR, and it didn't include the `log_xmllib_err` lines. So I removed them from `regression.tools.exp`. All CTS tests ran fine on my machine after that, and I pushed.

CI passed `build-all-voting` and failed `build-rpms`. I reproduced in my machine that `make -C rpm rpm` fails due to a diff in the `cts-cli` tests.

It's fixed now after I put those lines back in. It turns out there's a really strange issue with `cts-cli`, where it prints the `log_xmllib_err` lines with certain combinations of tests but not others. Likewise, another test group (`tools` IIRC) `unpack_resources` errors with some test group combinations and not with others. I filed https://projects.clusterlabs.org/T548 for this issue. The best bet is to run the full `cts-cli` suite, to get the same output as the regression.*.exp files and to update them when needed.

**Comment by nrwahl2** [comment]
Feature set bump?

File: lib/pacemaker/pcmk_output.c [hunk 6]
```diff
@@ -1864,6 +1864,50 @@ attribute_xml(pcmk__output_t *out, va_list args)
...
+{
+    const char *rule_id = va_arg(args, const char *);
+    int result = va_arg(args, int);
+
+    char *rc_str = pcmk__itoa(pcmk_rc2exitc(result));
+
+    pcmk__output_create_xml_node(out, "rule-check",
+                                 "rule-id", rule_id,
+                                 "rc", rc_str,
+                                 NULL);
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_output.c:1904]: The messages from rule_check_default could be added to the XML for additional clarity. Not sure if that's desirable and whether that would be allowable in the context of schema compatibility.

File: lib/pacemaker/pcmk_rule.c [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+                                  "and date_spec/@years "
+                                  "and not(date_spec/@moon)]", rule_id);
+        xpath_obj = xpath_search(cib_constraints, xpath);
+        max = numXpathResults(xpath_obj);
+
+        if (max == 0) {
+            rc = ENXIO;
+            out->err(out, "Rule either must not use date_spec, or use "
+                     "date_spec with years= but not moon=");
+            goto done;
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_rule.c]: We could potentially still call `out->message()` in this "unsupported" case, and add a handler to the format functions for `ENXIO`.

Maybe even do the same for the other cases where we bail out early (no rule found, more than one rule found, not exactly one date expression).

File: lib/pacemaker/pcmk_rule.c (hunk 1 above)
Comment (nrwahl2) [reply lib/pacemaker/pcmk_rule.c:241]: Updated to add static wrappers for calling with a single rule ID.

File: include/pacemaker.h (hunk 5 above)
Comment (kgaillot) [reply include/pacemaker.h:201]: Either is acceptable. Documenting in the header is easier for external developers who might think to look to the headers for documentation rather than the online API docs; documenting in the source is easier for us when reading code.

File: lib/pacemaker/pcmk_output.c (hunk 6 above)
Comment (kgaillot) [reply lib/pacemaker/pcmk_output.c:1904]: Not worth the effort at this point

File: lib/pacemaker/pcmk_rule.c [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+    }
+
+    pe_free_working_set(data_set);
+    return rc;
+}
+
+// Documented in pacemaker.h
+int
+pcmk_check_rules(xmlNodePtr *xml, xmlNodePtr input, crm_time_t *date,
+                 const char **rule_ids)
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_rule.c:273]: The public API should be added in its own commit for the message

File: tools/Makefile.am [hunk 9]
```diff
@@ -139,7 +139,8 @@ crm_rule_SOURCES 	= crm_rule.c
 crm_rule_LDADD		= $(top_builddir)/lib/cib/libcib.la		\
 			  $(top_builddir)/lib/pengine/libpe_rules.la   \
 			  $(top_builddir)/lib/pengine/libpe_status.la   \
-			  $(top_builddir)/lib/common/libcrmcommon.la
+			  $(top_builddir)/lib/common/libcrmcommon.la	\
+			  $(top_builddir)/lib/pacemaker/libpacemaker.la
```
Comment (kgaillot) [inline tools/Makefile.am:143]: libraries are listed highest-to-lowest level left-to-right (really only has an impact when linking statically, but it's good habit)

**Review by kgaillot — COMMENTED** [review]
Not a comprehensive review, just a first look

**Comment by kgaillot** [comment]
> Feature set bump?

No, since the tool interface isn't changing

File: tools/crm_rule.c (hunk 3 above)
Comment (kgaillot) [reply tools/crm_rule.c:248]: ? It's in include/crm/pengine/rules.h

Comment (nrwahl2) [reply tools/crm_rule.c:248]: Ouch, yeah thank you. I saw this below, with `time_expr` in the doxygen block of the helper function when the argument was really called `expr`, and blanked out. It wouldn't have compiled if `time_expr` weren't defined.
```
 * \param[in]  time_expr    date_expression XML
...
 */
static int
eval_date_expression(xmlNode *expr, crm_time_t *now, crm_time_t *next_change)
```

Will fix.

File: tools/Makefile.am (hunk 9 above)
Comment (nrwahl2) [reply tools/Makefile.am:143]: So libpacemaker first as the highest-level library? Not a pattern that's consistently followed so I wasn't aware of it. Easy switch.

File: lib/pacemaker/pcmk_rule.c (hunk 8 above)
Comment (nrwahl2) [reply lib/pacemaker/pcmk_rule.c:273]: Done

File: include/pacemaker.h [hunk 10]
```diff
@@ -186,6 +186,40 @@ int pcmk_list_nodes(xmlNodePtr *xml, char *node_types);
...
+ * \param[in]     input     The CIB XML to check (if \c NULL, use current CIB)
+ * \param[in]     date      Check whether the rule is in effect at this date and
+ *                          time (if \c NULL, use current date and time)
+ * \param[in]     rule_ids  The IDs of the rules to check, as a <tt>NULL</tt>-
+ *                          terminated list.
+ *
+ * \return Standard Pacemaker return code
+ */
+int pcmk_check_rules(xmlNodePtr *xml, xmlNodePtr input, crm_time_t *date,
+                     const char **rule_ids);
```
Comment (clumens) [inline include/pacemaker.h:202]: This function could also be written to take `...` for the list of rules.  I'd have to look at how it's used elsewhere, but it could potentially save some work at the callers since you wouldn't have to construct the data type to pass.

File: include/pcmki/pcmki_rule.h [hunk 11]
```diff
@@ -0,0 +1,31 @@
+#ifndef PCMK__PCMKI_PCMKI_RULE__H
```
Comment (clumens) [inline include/pcmki/pcmki_rule.h:1]: Remember the copyright boilerplate up here, too.

File: lib/pacemaker/pcmk_rule.c [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+        }
+    }
+
+    match = getXpathResult(xpath_obj, 0);
+
+    /* We should have ensured this with the xpath query above, but double-
+     * checking can't hurt.
+     */
+    CRM_ASSERT(match != NULL);
+
```
Comment (clumens) [inline lib/pacemaker/pcmk_rule.c:219]: You dropped the check of `find_expression_type` here.

File: lib/pacemaker/pcmk_rule.c (hunk 7 above)
Comment (clumens) [reply lib/pacemaker/pcmk_rule.c]: I don't really have any strong feelings on that one way or the other.  Maybe try converting one piece and see what it looks like.

File: include/pacemaker.h (hunk 10 above)
Comment (nrwahl2) [reply include/pacemaker.h:202]: It would be easier for some callers so I like it in theory. However, I'm not aware of a way to convert an existing array of strings to a list of separate arguments (or even a `va_list`), and searching isn't turning anything useful up. Closest search result I found said you can't: https://stackoverflow.com/a/14705987/7660197. Let me know if you have ideas that don't require having a variadic version and a non-variadic version and we can give it a shot

I figure at least accepting a char* array is clear and straightforward, even if it requires a little bit of extra typing sometimes.

File: lib/pacemaker/pcmk_rule.c [hunk 13]
```diff
@@ -126,15 +126,15 @@ rule_init(pcmk__output_t *out, xmlNodePtr input, crm_time_t *date,
  * \return Standard Pacemaker return code
  */
 static int
-crm_rule_check(pcmk__output_t *out, pe_working_set_t *data_set,
-               const char *rule_id, crm_time_t *date)
+do_check_rule(pcmk__output_t *out, pe_working_set_t *data_set,
```
Comment (clumens) [inline lib/pacemaker/pcmk_rule.c:129]: I know @kgaillot isn't really a fan of `do_` functions.

File: lib/pacemaker/pcmk_rule.c (hunk 12 above)
Comment (nrwahl2) [reply lib/pacemaker/pcmk_rule.c:219]: Thank you, yeah, that was a brain fart. Intentional at the time but definitely wrong: https://github.com/ClusterLabs/pacemaker/pull/2836#discussion_r957663996

File: lib/pacemaker/pcmk_rule.c (hunk 13 above)
Comment (nrwahl2) [reply lib/pacemaker/pcmk_rule.c:129]: Fair enough I thought `crm_rule_check` looked too much like a public name, and a bare `rule_check` or `check_rule` could be confusing when we already have `pcmk__rule_check` and `pcmk_rule_check` (the inline wrappers).

File: include/pcmki/pcmki_rule.h (hunk 11 above)
Comment (nrwahl2) [reply include/pcmki/pcmki_rule.h:1]: Thanks! I copied `include/pcmki/pcmki_cluster_queries.h` to use as my template, and it doesn't have copyright boilerplate. I'll add that before the next push.

File: lib/pacemaker/pcmk_rule.c [hunk 14]
```diff
@@ -155,58 +155,58 @@ crm_rule_check(pcmk__output_t *out, pe_working_set_t *data_set,
...
-        if (max == 0) {
-            rc = ENXIO;
+        free(xpath);
+
+        if (num_results == 0) {
             out->err(out, "Rule must either not use date_spec, or use "
                      "date_spec with years= but not moon=");
-            goto done;
+            freeXpathObject(xpath_obj);
+            return ENXIO;
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_rule.c:209]: Seems like `EOPNOTSUPP` might be better here too

Comment (nrwahl2) [reply lib/pacemaker/pcmk_rule.c]: Done; will revert if we're concerned about anyone depending on `ENXIO` specifically in this situation

File: lib/pacemaker/pcmk_rule.c (hunk 7 above)
Comment (nrwahl2) [reply lib/pacemaker/pcmk_rule.c]: Looks fine to me. I'll point it out in the new regression test output in a sec. I don't have strong feelings here either.

**Comment by nrwahl2** [comment]
Updated:
* Return `EOPNOTSUPP` instead of `ENXIO` when a rule contains a date_spec without `years` or with `years=` and `moon=`. This is something we don't support checking, not a rule not found situation.
* Added copyright to `pcmki_rule.h` and `pcmki_cluster_queries.h`.
* Separated public API additions into their own commit
* Moved libpacemaker to the beginning of `LDADD` for `crm_rule`
* Removed some redundant arguments in `eval_rule_expression()` and `do_check_rule()`. Use `data_set->now` after initialization so that we can drop an argument to those helpers and also make the `date` arguments const for the API functions.
* Re-added the `time_expr` line that I mistakenly dropped.
* Write a `"rule-check"` message when we fail to check a rule, so that we have a check result item for every rule that we request to check, even if we can't find it or don't support checking it.
* Added XML versions with validations to all `crm_rule` regression tests that were missing them.

I've got a small mistake in regression test output on this update, so I'm about to push again to fix that after `make -C rpm rpm` finishes on my machine to validate.

File: cts/cli/regression.rules.exp [hunk 15]
```diff
@@ -28,6 +28,7 @@ crm_rule: Couldn't parse input from STDIN
 * Passed: echo           - crm_rule given invalid input XML on stdin
 =#=#=#= Begin test: Try to check a rule that doesn't exist =#=#=#=
 No rule found with ID=blahblah
+Could not determine whether rule blahblah is expired
```
Comment (nrwahl2) [inline cts/cli/regression.rules.exp:79]: @clumens Example of new text output when we call `out->message` in the event of rule-check errors too.

File: cts/cli/regression.rules.exp [hunk 16]
```diff
@@ -91,6 +92,7 @@ No rule found with ID=blahblah
 * Passed: crm_rule       - Try to check a rule that doesn't exist
 =#=#=#= Begin test: Try to check a rule that doesn't exist, with XML output =#=#=#=
 <pacemaker-result api-version="X" request="crm_rule -c -r blahblah --output-as=xml">
+  <rule-check rule-id="blahblah" rc="105"/>
```
Comment (nrwahl2) [inline cts/cli/regression.rules.exp:143]: @clumens Example of new XML output when we call out->message in the event of rule-check errors too.

One advantage is that I think previously, the status code would be the last error we found. Now we get a discrete rule-check element and return code for every rule we checked.

File: include/pacemaker.h (hunk 10 above)
Comment (clumens) [reply include/pacemaker.h:202]: There's no way that I know of to do that.

File: cts/cli/regression.rules.exp [hunk 17]
```diff
@@ -145,13 +148,16 @@ Rule cli-prefer-rule-dummy-date_spec-only-years satisfies conditions
 * Passed: crm_rule       - Verify date_spec rule with years is in effect
 =#=#=#= Begin test: Try to check a rule whose date_spec does not contain years= =#=#=#=
 Rule must either not use date_spec, or use date_spec with years= but not moon=
+Could not determine whether rule cli-prefer-rule-dummy-date_spec-without-years is expired
```
Comment (clumens) [inline cts/cli/regression.rules.exp:238]: Looking at these two next to each other, it'd be nice if the top message also said "Rule cli-prefer-rule-dummy-date_spec-without-years must either...".

File: lib/pacemaker/pcmk_rule.c (hunk 13 above)
Comment (kgaillot) [reply lib/pacemaker/pcmk_rule.c:122]: Maybe `check_rule_effectiveness` or `eval_rule`

File: cts/cli/regression.rules.exp (hunk 16 above)
Comment (kgaillot) [reply cts/cli/regression.rules.exp:143]: That does seem helpful. Need to update the xml/api/crm_rule schema and bump PCMK__API_VERSION

File: cts/cli/regression.rules.exp (hunk 15 above)
Comment (kgaillot) [reply cts/cli/regression.rules.exp:79]: The text seems redundant, not sure what we should do here. The ideal would be something like

    Could not determine whether rule is expired: Rule blahblah not found

but that's probably not worth the hassle of figuring out

File: lib/pacemaker/pcmk_rule.c [hunk 18]
```diff
@@ -152,11 +152,13 @@ do_check_rule(pcmk__output_t *out, pe_working_set_t *data_set,
...
-    } else if (num_results > 1) {
-        out->err(out, "More than one rule with ID=%s found", rule_id);
+    if (num_results != 1) {
+        if (num_results == 0) {
+            out->err(out, "No rule found with ID=%s", rule_id);
+        } else {    // num_results > 1
+            out->err(out, "More than one rule with ID=%s found", rule_id);
+        }
+        out->message(out, "rule-check", rule_id, ENXIO);
         return ENXIO;
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_rule.c:162]: We have pcmk_rc_multiple but unfortunately that's specifically for "Resource active on multiple nodes". I think this sort of thing comes up enough to be worth a more generic code like pcmk_rc_duplicate_id.

**Review by kgaillot — COMMENTED** [review]
nice work

File: cts/cli/regression.rules.exp (hunk 16 above)
Comment (nrwahl2) [reply cts/cli/regression.rules.exp:143]: zeroOrMore to oneOrMore rule-checks? Can't think of what else might need to change

File: cts/cli/regression.rules.exp (hunk 17 above)
Comment (nrwahl2) [reply cts/cli/regression.rules.exp:238]: Good point. The other three "failed to check rule" messages all include the rule ID.

File: lib/pacemaker/pcmk_output.c [hunk 19]
```diff
@@ -1864,6 +1864,50 @@ attribute_xml(pcmk__output_t *out, va_list args)
...
+    } else if (result == pcmk_rc_after_range) {
+        out->info(out, "Rule %s is expired", rule_id);
+    } else if (result == pcmk_rc_before_range) {
+        out->info(out, "Rule %s has not yet taken effect", rule_id);
+    } else if (result == pcmk_rc_op_unsatisfied) {
+        out->info(out, "Rule %s does not satisfy conditions", rule_id);
+    } else {
+        out->info(out, "Could not determine whether rule %s is expired",
+                  rule_id);
+    }
```
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/pacemaker/pcmk_output.c:1887]: Yep, will do

File: lib/pacemaker/pcmk_rule.c [hunk 20]
```diff
@@ -0,0 +1,272 @@
...
+    }
+
+    pe__set_working_set_flags(new_data_set,
+                              pe_flag_no_counts|pe_flag_no_compat);
+
+    // Populate the working set instance
+
+    // Make our own copy of the given input or fetch the CIB and use that
+    if (input != NULL) {
+        new_data_set->input = copy_xml(input);
```
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/pacemaker/pcmk_rule.c:80]: I noticed that, starting to look through the libxml2 docs recently... kind of frustrating for functions that clearly don't modify anything

File: cts/cli/regression.rules.exp (hunk 16 above)
Comment (kgaillot) [reply cts/cli/regression.rules.exp:143]: New schemas should remain compatible with output produced by earlier versions, so zeroOrMore

File: lib/pacemaker/pcmk_rule.c [hunk 21]
```diff
@@ -0,0 +1,272 @@
...
+{
+    xmlNodePtr cib_constraints = NULL;
+    xmlNodePtr match = NULL;
+    xmlXPathObjectPtr xpath_obj = NULL;
+    char *xpath = NULL;
+    int rc = pcmk_rc_ok;
+    int max = 0;
+
+    CRM_ASSERT((out != NULL) && (data_set != NULL) && (rule_id != NULL)
+               && (date != NULL));
```
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/pacemaker/pcmk_rule.c]: Fair enough, I'm just paranoid. Will remove it.

File: lib/pacemaker/pcmk_rule.c [hunk 22]
```diff
@@ -0,0 +1,272 @@
...
+     * to check. For the moment, these rules must only have a single
+     * date_expression child and:
+     * - Do not have a date_spec operation, or
+     * - Have a date_spec operation that contains years= but does not contain
+     *   moon=.
+     *
+     * We do this in steps to provide better error messages. First, check that
+     * there's any rule with the given ID.
+     */
+    xpath = crm_strdup_printf("//rule[@id='%s']", rule_id);
```
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/pacemaker/pcmk_rule.c:148]: Oh yeah I'd prefer that. I didn't think to check for it.

File: cts/cli/regression.rules.exp (hunk 16 above)
Comment (nrwahl2) [reply cts/cli/regression.rules.exp:143]: I'm saying I don't see what in the schema needs to be updated for this PR. (There is every chance in the world I've got a glaring blind spot right now.)

Comment (kgaillot) [reply cts/cli/regression.rules.exp:143]: My bad, I thought this was completely new but it's just a new occurrence

Comment (nrwahl2) [reply cts/cli/regression.rules.exp:143]: Ah gotcha, no worries

File: lib/pacemaker/pcmk_rule.c (hunk 13 above)
Comment (nrwahl2) [reply lib/pacemaker/pcmk_rule.c:122]: `eval_rule` seems like a good fit since `eval` is the pattern for this kind of thing in `rules.c` (e.g., `pe_eval_rules`, `pe_eval_expr`). It's almost too generic since we might want to use `eval_rule` in something like `rules.c` later -- but it's static so we can change it if we ever need to. (Static name conflicts won't break compilation, just good to avoid for logging purposes).

File: include/pacemaker.h (hunk 10 above)
Comment (nrwahl2) [reply include/pacemaker.h:202]: Let me know; we can do it either way. I think there's a reasonable possibility that a caller will want to check multiple rules at once and parse the results together. Maybe fairly uncommon though...

I'd be less concerned about efficiency since it's probably an uncommon need and maybe not super time-sensitive; and more concerned about clunkiness for a caller. I've been annoyed at REST APIs that only let me get a single item at a time before.

Comment (kgaillot) [reply include/pacemaker.h:202]: Might as well leave it as it is

File: cts/cli/regression.rules.exp (hunk 15 above)
Comment (nrwahl2) [reply cts/cli/regression.rules.exp:79]: I think it can be done without too much trouble, if we want to send the "Could not determine" message to `out->err` instead of `out->info` like we do with the other formatter messages.

Comment (kgaillot) [reply cts/cli/regression.rules.exp:79]: Seems like it should be an error anyway

Comment (nrwahl2) [reply cts/cli/regression.rules.exp:79]: @kgaillot I've got it working so that it all prints on one line for "rule not found" and "more than one rule" cases for text output. The unsupported date format case would still have two lines, because we'd be returning `EOPNOTSUPP` for both subcases and the formatter has no way to differentiate.

The only caveat is that the error **messages** for "rule not found" and "more than one rule" won't show up in XML output. XML output would only get the return code. (XML output would still get the error messages for the unsupported date formats.)

That might be okay, as long as we use each return code for a unique situation: `ENXIO` for "rule not found" and `pcmk_rc_duplicate_id` for "more than one rule." I'm doing that in the next update regardless.

Thoughts?

```
Could not determine whether rule blahblah is expired: Rule not found
...
Could not determine whether rule duplicate-rule-id is expired: Found more than one rule with matching ID
...
Rule cli-prefer-rule-dummy-date_spec-years-moon must either not use date_spec, or use date_spec with years= but not moon=
Could not determine whether rule cli-prefer-rule-dummy-date_spec-years-moon is expired: Unsupported date format
...
Can't check rule cli-no-date_expression-rule because it does not have exactly one date_expression
Could not determine whether rule cli-no-date_expression-rule is expired: Unsupported date format
```

```
<pacemaker-result api-version="X" request="crm_rule -c -r blahblah --output-as=xml">
  <rule-check rule-id="blahblah" rc="105"/>
  <status code="105" message="No such object">
  <!-- this would go away
    <errors>
      <error>No rule found with ID=blahblah</error>
    </errors>
  -->
  </status>
</pacemaker-result>
```

File: lib/pacemaker/pcmk_rule.c (hunk 18 above)
Comment (nrwahl2) [reply lib/pacemaker/pcmk_rule.c:162]: agreed, done

File: cts/cli/regression.rules.exp (hunk 15 above)
Comment (nrwahl2) [reply cts/cli/regression.rules.exp:79]: I'll go ahead and push it with the other changes. It's easy to modify if we decide we don't want this.

**Comment by nrwahl2** [comment]
* Rename do_check_rule() to eval_rule(), rule_init to init_rule_check()
* Use `XML_TAG_RULE`
* Add rule ID to `"moon="` error message, like the others
* Remove asserts from static functions
* Use switch in `rule_check_default()` and return the value of `out->info()` to make it compact
* Add `pcmk_rc_duplicate_id` and use it when more than one matching rule is found
* Put "rule not found" and "more than one matching rule found" messages on same line as "Could not determine..." (but drop them from XML output) (up for discussion of whether we want this since it drops "no rule" and "multiple rules" from XML error output -- https://github.com/ClusterLabs/pacemaker/pull/2836#discussion_r958952687)

File: lib/pacemaker/pcmk_rule.c [hunk 23]
```diff
@@ -111,6 +111,8 @@ init_rule_check(pcmk__output_t *out, xmlNodePtr input, crm_time_t *date,
     return pcmk_rc_ok;
 }
 
+#define XPATH_NODE_RULE "//" XML_TAG_RULE "[@id='%s']"
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_rule.c:109]: Some of the other xpath components have constants e.g. XML_ATTR_ID, XML_EXPR_ATTR_OPERATION

File: include/crm/common/results.h [hunk 24]
```diff
@@ -108,6 +108,7 @@ enum pcmk_rc_e {
     /* When adding new values, use consecutively lower numbers, update the array
      * in lib/common/results.c, and test with crm_error.
      */
+    pcmk_rc_duplicate_id        = -1033,
```
Comment (kgaillot) [inline include/crm/common/results.h:111]: This commit message would be API:

File: lib/pacemaker/pcmk_output.c [hunk 25]
```diff
@@ -1884,10 +1884,30 @@ rule_check_default(pcmk__output_t *out, va_list args)
...
-                             "Could not determine whether rule %s is expired",
-                             rule_id);
+            break;
+    }
+
+    /* text_err() prints to stderr, hard-coded. We can't call out->err() here
+     * because we don't want a newline.
+     */
+    fprintf(stderr, "Could not determine whether rule %s is expired: ",
+            rule_id);
```
Comment (kgaillot) [inline lib/pacemaker/pcmk_output.c:1894]: For future-proofing (vs new or reimplemented formats) I'd rather just duplicate this in the out->err() calls. That also lets us merge the switches and simplify the default

File: cts/cli/regression.rules.exp [hunk 26]
```diff
@@ -91,16 +91,14 @@ No rule found with ID=blahblah
 * Passed: crm_rule       - Try to check a rule that doesn't exist
 =#=#=#= Begin test: Try to check a rule that doesn't exist, with XML output =#=#=#=
 <pacemaker-result api-version="X" request="crm_rule -c -r blahblah --output-as=xml">
-  <status code="105" message="No such object">
-    <errors>
-      <error>No rule found with ID=blahblah</error>
-    </errors>
-  </status>
```
Comment (kgaillot) [inline cts/cli/regression.rules.exp:98]: I hate losing detail in the XML format because we're pushing everyone to use it, and I don't want to give them a reason not to. We can call the same out->err()'s in the XML implementation

**Review by kgaillot — COMMENTED** [review]
These are nice improvements

File: include/crm/common/results.h (hunk 24 above)
Comment (nrwahl2) [reply include/crm/common/results.h:111]: Makes sense, wish I'd caught that in 73a46d1a :)
Will fix

File: lib/pacemaker/pcmk_rule.c (hunk 23 above)
Comment (nrwahl2) [reply lib/pacemaker/pcmk_rule.c:109]: I must have stopped checking after I saw that `date_expression` didn't have a constant

File: lib/pacemaker/pcmk_output.c (hunk 25 above)
Comment (nrwahl2) [reply lib/pacemaker/pcmk_output.c:1894]: Good point

File: cts/cli/regression.rules.exp (hunk 26 above)
Comment (nrwahl2) [reply cts/cli/regression.rules.exp:98]: Good point. At that point I might as well move the `out->err()` calls back to `eval_rule()` in `pcmk_rule.c` so that they printed for all formats. Then remove the default error from the `rule_check_default` formatter.

**Comment by nrwahl2** [comment]
Updated to address review.
* Use `XML_ATTR_ID` and `XML_EXPR_ATTR_OPERATION`. I don't see any more that have defined constants.
* Update `pcmk_rc_duplicate_id` commit message to "API:"
* Re-did the error handling again. I appreciate the push to reorganize it. Whether we stick with this approach or not, I like it a lot better.
  * Now `eval_rule()` doesn't have any side effects. We just pass the rc and error message (via pointer) back to the caller let the caller call `out->message()`.
  * `out->message()` now accepts an error argument. The default case uses the error if it's not `NULL`, or "unexpected error" if it is `NULL`.
  * The XML formatter also calls `out->err()` in the event of an error.

File: lib/pacemaker/pcmk_output.c [hunk 27]
```diff
@@ -1864,6 +1864,65 @@ attribute_xml(pcmk__output_t *out, va_list args)
...
+{
+    const char *rule_id = va_arg(args, const char *);
+    int result = va_arg(args, int);
+    const char *error = va_arg(args, const char *);
+
+    switch (result) {
+        case pcmk_rc_within_range:
+            return out->info(out, "Rule %s is still in effect", rule_id);
+        case pcmk_rc_ok:
+            return out->info(out, "Rule %s satisfies conditions", rule_id);
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_output.c:1879]: I don't think this is very intuitive... for text output, I wonder if should we combine both `pcmk_rc_within_range` and `pcmk_rc_ok` into one case for "Rule %s is in effect". Or something like `"Rule %s is in effect: within date range"` and `"Rule %s is in effect: satisfies cron conditions"`

File: lib/pacemaker/pcmk_output.c (hunk 6 above)
Comment (nrwahl2) [reply lib/pacemaker/pcmk_output.c:1908]: With the latest update, this seems trivial to implement if we wanted to do so. We could add an optional `message` attribute and set it to a known message or something like `"Error occurred checking rule"`. It could go in a separate PR if we wanted it, since this one is already unwieldy.

File: lib/pacemaker/pcmk_output.c [hunk 28]
```diff
@@ -1864,6 +1864,65 @@ attribute_xml(pcmk__output_t *out, va_list args)
...
+
+PCMK__OUTPUT_ARGS("rule-check", "const char *", "int", "const char *")
+static int
+rule_check_xml(pcmk__output_t *out, va_list args)
+{
+    const char *rule_id = va_arg(args, const char *);
+    int result = va_arg(args, int);
+    const char *error = va_arg(args, const char *);
+
+    char *rc_str = pcmk__itoa(pcmk_rc2exitc(result));
```
Comment (nrwahl2) [inline lib/pacemaker/pcmk_output.c:1903]: I wonder if we should map both success codes (`pcmk_rc_ok` and `pcmk_rc_within_range`) to `pcmk_rc_ok`, to make things simpler for the caller. An rc of 0 would mean "the rule is in effect" -- instead of checking for two different codes. I don't think it would deprive anyone of useful information.

Not sure if we really want `pcmk_rc2exitc` here but I'm leaving it alone unless requested.

File: lib/pacemaker/pcmk_output.c (hunk 27 above)
Comment (kgaillot) [reply lib/pacemaker/pcmk_output.c:1879]: I wouldn't bother at this point

File: lib/pacemaker/pcmk_rule.c [hunk 29]
```diff
@@ -0,0 +1,296 @@
...
+    free(xpath);
+    freeXpathObject(xpath_obj);
+
+    if (num_results == 0) {
+        *error = "Rule not found";
+        return ENXIO;
+    }
+
+    if (num_results > 1) {
+        // Should not be possible; schema prevents this
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_rule.c:160]: One gotcha that's easy to forget is that schema validation can be disabled, and then the input can be anything as long as it's legal XML. So while the schema validation is important for the user experience, it can't be relied on by the code.

**Review by kgaillot — APPROVED** [review]
Stick a fork in it!
